rust/src/qfunction.rs

```diff
@@ -392,14 +392,14 @@ impl QFunction {
     /// qf = qf.input("u", 1, EvalMode::Interp);
     /// qf = qf.input("weights", 1, EvalMode::Weight);
     /// ```
-    pub fn input(mut self, fieldname: &str, size: i32, emode: crate::EvalMode) -> Self {
+    pub fn input(mut self, fieldname: &str, size: usize, emode: crate::EvalMode) -> Self {
         let name_c = CString::new(fieldname).expect("CString::new failed");
         let idx = self.trampoline_data.number_inputs;
         self.trampoline_data.input_sizes[idx] = size;
         self.trampoline_data.number_inputs += 1;
         let emode = emode as bind_ceed::CeedEvalMode;
         unsafe {
-            bind_ceed::CeedQFunctionAddInput(self.qf_core.ptr, name_c.as_ptr(), size, emode);
+            bind_ceed::CeedQFunctionAddInput(self.qf_core.ptr, name_c.as_ptr(), size as i32, emode);
```

Comment:
You shouldn't cast `size` to `i32` inside an `unsafe` block.

Reply:
why not?

Reply:
because casting can fail

Reply:
Casting between a usize and an i32 shouldn't be able to fail

Reply:
`usize` is "bigger" than `i32` no?

Reply:
(the max is not the same)

Reply:
The cast will truncate and will always be successful, but we can put in a check that the value is less than the max i32 I guess

Reply:
We want it to panic if it's out of range, do it unsafely and check manually. Presumably you can bind it outside.

We could consider updating the libCEED interface to use `size_t` in more places. The only place where there is a performance impact of size is indexing in `ElemRestriction`. Otherwise, we used signed ints mainly because they avoid possible bugs in reverse `for (i=n; i>=0; i--)` loops, though an equally valid point is that negative values should not exist in certain circumstances and the type system doesn't help us.

Reply:
Integer overflow only panics in debugging mode, not in release, so we'd need to add the check manually if we want it to always occur (which would be my preference)

Reply:
Isn't it what `try_from` does?

Reply:
I don't think we can consider that it is a success that our value gets truncated.

Reply:
I think we just agreed to check for overflow

Reply:
Sorry, I didn't update the page before sending my messages.

Reply:
GitHub usually automatically refreshes for me, so I'm not sure why it isn't today, but I'm having to manually refresh as well

Reply:
Again, isn't it what `try_from` does?

Reply:
@jedbrown What's the benefit of doing it unsafely and checking manually?

Reply:
Something like:
```rust
let size = i32::try_from(size)?;
unsafe { ... }
```

Reply:
`try_from` is good. I didn't see the context of whether we were in a `Result`. We want it checked (necessarily in debug mode, but this is sufficiently coarse granularity that checking always won't be perceptible in the performance profile.